### PR TITLE
feat: introduce rank-local and synced source of randomness

### PR DESCRIPTION
**Pull request by ssmmnn11**
Introduce separate model RNG streams for distributed training. The default pytorch RNG is now seeded independently per rank, so dropout can use the rank-local random source. A synced pytorch RNG stream is available through a context manager for operations that must stay identical across ranks, such as model initialization and selected noise sampling operations.

### CODE REVIEW

**Comment by ssmmnn11** [comment]
#595

File: models/src/anemoi/models/layers/ensemble.py [hunk 1]
```diff
@@ -181,7 +182,8 @@ def forward(
             self.noise_channels,
         )
```
Comment (VeraChristina) [inline models/src/anemoi/models/layers/ensemble.py:192]: Can we make sure that this is the same across all ranks, specifically for grid_size if self.noise_graph_provider is None? I.e. prevent callers from passing shard-local sizes which would desyncronize the synced stream. E.g. take the full grid size and shard inside, or add a check?

File: models/src/anemoi/models/layers/ensemble.py
Comment (VeraChristina) [inline models/src/anemoi/models/layers/ensemble.py:218]: Do we need maybe_checkpoint here?

File: training/src/anemoi/training/train/train.py [hunk 2]
```diff
@@ -348,7 +350,9 @@ def model(self) -> pl.LightningModule:
 
         training_method_cfg = self.config.training.method
         training_method_cls = get_class(training_method_cfg._target_)
-        model = instantiate_with_runtime_kwargs(training_method_cfg, **kwargs)  # Task -> pl.LightningModule
+        seed_torch_rng_sources(self.initial_seed, self.strategy.global_rank, reset_synced=True)
```
Comment (VeraChristina) [inline training/src/anemoi/training/train/train.py:370]: This resets the groups to the stream derived from sync_group_id=0. Elsewhere, we set sync_group_id=model_comm_group_id. Do we need to restore this after model init?

File: models/src/anemoi/models/distributed/random.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+    new_cuda_devices = _cuda_devices_in_state(current_state).difference(
+        _cuda_devices_in_state(default_state),
+    )
+    if new_cuda_devices:
+        msg = (
+            "CUDA was initialized inside use_synced_torch_rng(). "
+            "Initialize CUDA before entering the synced random context."
+        )
+        raise RuntimeError(msg)
```
Comment (VeraChristina) [inline models/src/anemoi/models/distributed/random.py:295]: This raises after modifying the _synced_state. This seems fine for training runs because they crash here, but not sure how it interacts with tests.

File: models/tests/layers/processor/test_pointwise_processor.py [hunk 4]
```diff
@@ -71,15 +71,71 @@ def test_pointwisemlp_processor_with_sharding_dropout_forward(pointwisemlp_proce
...
+        x,
+        batch_size,
+        shard_info,
+        model_comm_group=fake_model_comm_group,
+    )
+
+    assert output.shape == x.shape
+
+
+def test_pointwisemlp_processor_checkpointed_dropout_preserves_rng(pointwisemlp_processor_init):
```
Comment (VeraChristina) [inline models/tests/layers/processor/test_pointwise_processor.py:86]: do we need a seed_torch_rng_sources here?

**Review by VeraChristina — COMMENTED** [review]
Thanks for adding this! I think the separate saved RNG state acting as synced stream generally works well. A couple of comments below mostly regarding maintainability, like making sure future contributions don't desyncronise the synced steam.

One small comment along those lines regarding RandomFourierEmbeddigns in diffusion.py (which wasn't modified so I can't comment there directly): I would prefer the torch.randn call in the init to be explicitly wrapped in use_synced_torch_rng and not just implicitly being called in the model construction in train.py (which is wrapped in the context) -- just to be more explicit, it works as it is.

More generally, I am wondering if we can make the two states of the stream more explicit for future contributors, e.g. having `with_synced_torch_rng` and `with_local_torch_rng`. (This doesn't enforce the correct context but it might help against forgetting about the context.) Maybe even wrap torch.randn calls in `synced_randn` and `local_randn` which assert they are in the correct context?

Similarly, perhaps it's worth thinking about adding guards against rank-specific calls to toch.randn within synced contexts? E.g. an optional check for shape consistency (broadcast shape from rank 0 and assert all ranks agree) inside synced_randn controlled by a debug flag? Then we could add an integration test that runs in debug mode. (Maybe this is too cautious but I imagine these type of issues would be hard to debug.)
